### PR TITLE
KAFKA-2574: Add ducktape based ssl test for KafkaLog4jAppender

### PR DESCRIPTION
**Pull request by SinghAsDev**

### CODE REVIEW

**Comment by SinghAsDev** [comment]
Note: this needs #175

**Comment by SinghAsDev** [comment]
@gwenshap this should address your concern on KAFKA-2447. Could you take a look at this and #175.

**Comment by granders** [comment]
@SinghAsDev This seems to be failing when run with SSL:
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/108/artifact/results/KAFKA-2574/latest/report.html

**Comment by SinghAsDev** [comment]
@granders did you cherry-pick #175 before running this?

**Comment by granders** [comment]
@SinghAsDev Sorry I didn't see you had broken this into two PRs. We'll probably want to merge in #175 first so that we're not adding failing tests.

**Comment by SinghAsDev** [comment]
No problem. I did mention on the PR, but github comments are hard to follow. Definitely #175 should go in before this, this PR was created to address @gwenshap concern on #175 .

**Comment by gwenshap** [comment]
@granders I pushed #175 to trunk. Do you have other comments on this, or is it good to go?

File: tools/src/main/java/org/apache/kafka/clients/tools/VerifiableLog4jAppender.java [hunk 1]
```diff
@@ -88,6 +88,32 @@ private static ArgumentParser argParser() {
             .metavar("ACKS")
             .help("Acks required on each produced message. See Kafka docs on request.required.acks for details.");
 
+        parser.addArgument("--security-protocol")
```
Comment (granders) [inline tools/src/main/java/org/apache/kafka/clients/tools/VerifiableLog4jAppender.java]: I think mostly the tools have been extended for security by taking a config file, e.g. `--producer.config` (see for example `VerifiableProducer`) instead of several command line options

I think this approach will be a little easier to extend to sasl as well

File: tests/kafkatest/tests/log4j_appender_test.py [hunk 2]
```diff
@@ -25,27 +28,50 @@
...
         Tests if KafkaLog4jAppender is producing to Kafka topic
         :return: None
         """
-        self.appender.start()
+        self.start_kafka(security_protocol, security_protocol)
+        self.start_appender(security_protocol)
         self.appender.wait()
 
+        self.start_consumer(security_protocol)
         t0 = time.time()
```
Comment (granders) [inline tests/kafkatest/tests/log4j_appender_test.py]: Probably not necessary to report consumer startup time in this test? Also `start_consumer` is a misleading name, since it doesn't actually start the consumer :)

Also - on line 88, change `self.consumer.stop_node(node)` -> `self.consumer.stop()`

**Comment by granders** [comment]
@SinghAsDev @gwenshap A couple minor comments to address, otherwise looks good.
It would be great to rebase this so we can confirm that the tests are passing.

File: tests/kafkatest/tests/log4j_appender_test.py (hunk 2 above)
Comment (SinghAsDev) [inline tests/kafkatest/tests/log4j_appender_test.py]: Adopted.

File: tools/src/main/java/org/apache/kafka/clients/tools/VerifiableLog4jAppender.java (hunk 1 above)
Comment (SinghAsDev) [inline tools/src/main/java/org/apache/kafka/clients/tools/VerifiableLog4jAppender.java]: The problem is unlike `VerifiableProducer`, `VerifiableLog4jAppender` has different keys for these properties. So, one can not just append `SecurityConfig` to its property file. However, I do like the idea of being able to pass a property file and have added that as an option. Unfortunately, it does not buy us anything to pass properties to `VerifiableLog4jAppender` via a property file though.

**Comment by gwenshap** [comment]
Something went wrong with the PR... it is now showing 78 commits in 448 files

**Comment by SinghAsDev** [comment]
@gwenshap  fixed :)

**Comment by gwenshap** [comment]
Do I need to set up anything special to run the tests?

The PLAINTEXT tests pass, but SSL fails with  "Timed out waiting to consume expected number of messages."

**Comment by SinghAsDev** [comment]
Weird, nothing special should be needed. Trying to reproduce.

On Fri, Oct 30, 2015 at 3:53 PM, Gwen (Chen) Shapira <
notifications@github.com> wrote:

> Do I need to set up anything special to run the tests?
> 
> The PLAINTEXT tests pass, but SSL fails with "Timed out waiting to consume
> expected number of messages."
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/apache/kafka/pull/319#issuecomment-152669109.

## 

Regards,
Ashish

**Comment by SinghAsDev** [comment]
I think I know the issue, testing it out. This will require a HOTFIX, look
for PR shortly.

On Fri, Oct 30, 2015 at 4:06 PM, Ashish Singh asingh@cloudera.com wrote:

> Weird, nothing special should be needed. Trying to reproduce.
> 
> On Fri, Oct 30, 2015 at 3:53 PM, Gwen (Chen) Shapira <
> notifications@github.com> wrote:
> 
> > Do I need to set up anything special to run the tests?
> > 
> > The PLAINTEXT tests pass, but SSL fails with "Timed out waiting to
> > consume expected number of messages."
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/apache/kafka/pull/319#issuecomment-152669109.
> 
> ## 
> 
> Regards,
> Ashish

## 

Regards,
Ashish

**Comment by SinghAsDev** [comment]
@gwenshap after #395 gets committed, it should work fine. https://github.com/apache/kafka/commit/1cc44830b90688f1a2034243b8ee03b97101f75c got in between I rebased my changes and you tried this.

**Comment by gwenshap** [comment]
LGTM. New tests pass.
